### DISCUSSION TITLE
Hard limits behavior

### DISCUSSION THREAD

**Discussion by mvackel** [Paciente8159/uCNC#504] [Ideas]
(Using uCNC 1.7.4 with Mega/Ramps 1.4 and Reprapdiscount full graphic smart controller)

I'm testing the limit switches and hard limit set ($21). 

First, I can't directly change the hard limit using the display. The setting always returns to zero. I have to use UGS to change it and then use the display to save the setting (UGS doesn't save to EEPROM - probably I should change something in uCNC...).

With hard limit set, when any limit switch is triggered, the display freezes. UGS continues to behave normally: it issues an "Alarm" warning and locks, but the machine can continue with a Soft Reset and the display returns to operation.

I think the display should behave like UGS, issuing an Alarm in the main page and allowing some command to unlock the machine again. Am I correct?
Tested on 3-Axis and 5-axis configurations (configuration files from the web config). Curiously, the 5-axis shows the "Alarm", but also freezes.

**Comment by Paciente8159** [comment]
Currently, after an alarm the display is only updated. On Alarm condition system menu ceases to accept commands.

I will review that to add some sort of user unlock/soft reset screen.

**Comment by Paciente8159** [comment]
> First, I can't directly change the hard limit using the display.

Do have extended commands enabled? I so that might be the issue. You make the modifications but the $SS command is not being send to be permanently stored in EEPROM and on power off/on the changes are lost.

**EDIT**
Actually that is not the problem at all. On the system menu you need to explicitly store the settings or else they will remain in RAM only (or at least until store command is issued). In the case of the display this must be explicitly done.

**Comment by Paciente8159** [comment]
About the alarm/freeze issue...I've found a working solution.
Tomorrow I will upload the new changes.

**Comment by Paciente8159** [comment]
Added soft reset via display controller/encoder. Just keep the knob pressed for at least 5 seconds.
Theses were the changes
https://github.com/Paciente8159/uCNC-modules/pull/31
https://github.com/Paciente8159/uCNC/pull/507

**Reply by mvackel** [reply]
I've teste this and it works but I'd like to suggest, maybe in addiction to this approach, to include a menu command in the main menu (to be similar with UGS) that would have the present "Hold", "Resume", "Unlock" ... and a additional "Soft Reset" command. Also, the first page would display the Status "St: Alarm" as sometimes I had observed (don't know exactly the conditions) and, preferably, the alarmed switches (i.e.: "X"  or "Z"...). 
And, of course, do not freeze the display after the alarm, so you could navigate to the main menu and issue "Soft Reset"  and "Unlock" like we do in UGS.
This would make the Hard limits behavior similar to UGS, being more coherent and easy to understand.

**Comment by Paciente8159** [comment]
Inside the cnc_run function there are several smaller loops that can be divided in 3 stages. The unlock loop, the run loop and the reset loop.

When you enter an alarm condition you get out of the run loop and enter the reset loop that evaluates if a soft reset is required or not depending on the alarm type, before reentering the unlock loop.

On both reset and unlock loop only a subset of actions are possible (being the reset loop more restrictive).
There is no point in accessing the whole menu on the reset loop because you can't also do any thing via the serial communication commands. This restrict the user to check the system and have to do a soft reset and preventing any kind of action that might cause any undesired accident.

I understand what you said though. The popup alarm covers the window and you cannot see what triggered the alarm. I will redraw the alarm screen to provide better info on the possible crash reasons (like display limits, controls, alarm code, etc...)

**Comment by Paciente8159** [comment]
This is the redesign of the alarm screen with more info about the alarm.

https://github.com/Paciente8159/uCNC/pull/508
https://github.com/Paciente8159/uCNC-modules/pull/32

I'm just waiting on the checkups to merge the PR's.

Here is a sample
![image](https://github.com/Paciente8159/uCNC/assets/12116536/14397732-b094-4a2b-ab51-85eadefe3727)

I will start looking at the JOG with display issue.

**Reply by mvackel** [reply]
What is the "Alarm 13" message? I've observed various different numbers when testing.

Also, sometimes the 5s reset makes another alarm message to appear, requiring another reset - and this time the reset doesn't work. I'm trying to make a repeatable use case.

**Reply by Paciente8159** [reply]
The codes are the alarm codes that are original for Grbl for compatibility with existing software.
Some codes are new and are custom to µCNC.

The list of alarm codes can be checked [here](https://github.com/Paciente8159/uCNC/wiki/Basic-user-guide#alarms) and [here](https://github.com/Paciente8159/uCNC/blob/fa7ba3d2fd6638d7ba5dcc2208ae6eb968b4b902/uCNC/src/interface/grbl_interface.h#L140)

I'll see if I find any odd behavior on reset.
